File: library/src/main/java/su/levenetc/android/textsurface/animations/ShapeReveal.java
package su.levenetc.android.textsurface.animations;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.animation.FastOutSlowInInterpolator;

import su.levenetc.android.textsurface.Text;
import su.levenetc.android.textsurface.TextSurface;
import su.levenetc.android.textsurface.utils.Utils;
import su.levenetc.android.textsurface.interfaces.IEndListener;
import su.levenetc.android.textsurface.interfaces.ISurfaceAnimation;
import su.levenetc.android.textsurface.interfaces.ITextEffect;

/**
 * Created by Eugene Levenetc.
 */
public class ShapeReveal implements ITextEffect, ValueAnimator.AnimatorUpdateListener {

	private final Text text;
	private final int duration;
	private TextSurface textSurface;
	private IRevealShape revealShape;
	private boolean hideOnEnd;
	private ValueAnimator animator;

	public static ShapeReveal create(Text text, int duration, IRevealShape revealShape, boolean hideOnEnd) {
		return new ShapeReveal(text, duration, revealShape, hideOnEnd);
	}

	private ShapeReveal(Text text, int duration, IRevealShape revealShape, boolean hideOnEnd) {
		this.text = text;
		this.duration = duration;
		this.revealShape = revealShape;
		this.hideOnEnd = hideOnEnd;
		revealShape.setText(text);
	}

	@Override public void apply(Canvas canvas, String textValue, float x, float y, Paint paint) {
		revealShape.clip(canvas, textValue, x, y, paint);
	}

	@Override public void setInitValues(@NonNull Text text) {
		if (revealShape.isToShow()) text.setAlpha(0);
	}

	@NonNull @Override public Text getText() {
		return text;
	}

	@Override public void onStart() {
		text.addEffect(this);
	}

	@Override public void start(@Nullable final IEndListener listener) {

		text.setAlpha(255);
		animator = revealShape.getAnimator();
		animator.setInterpolator(new FastOutSlowInInterpolator());
		Utils.addEndListener(this, animator, new IEndListener() {
			@Override public void onAnimationEnd(ISurfaceAnimation animation) {
				text.removeEffect(ShapeReveal.this);
				if (hideOnEnd) text.setAlpha(0);
				if (listener != null) listener.onAnimationEnd(ShapeReveal.this);
			}
		});
		animator.setDuration(duration);
		animator.start();

	}

	@Override public void setTextSurface(@NonNull TextSurface textSurface) {
		revealShape.setTextSurface(textSurface);
		this.textSurface = textSurface;
	}

	@Override public long getDuration() {
		return 0;
	}

	@Override public void cancel() {
		if (animator != null && animator.isRunning()) {
			animator.cancel();
			animator = null;
		}
	}

	@Override public void onAnimationUpdate(ValueAnimator animation) {
		textSurface.invalidate();
	}

	public interface IRevealShape {
		void setText(Text text);

		void setTextSurface(TextSurface textSurface);

		ValueAnimator getAnimator();

		void clip(Canvas canvas, String textValue, float x, float y, Paint paint);

		boolean isToShow();
	}
}

File: library/src/main/java/su/levenetc/android/textsurface/animations/reveal/ShapeReveal.kt
package su.levenetc.android.textsurface.animations.reveal

import android.animation.ValueAnimator
import android.animation.ValueAnimator.AnimatorUpdateListener
import android.graphics.Canvas
import android.graphics.Paint
import androidx.interpolator.view.animation.FastOutSlowInInterpolator
import su.levenetc.android.textsurface.Text
import su.levenetc.android.textsurface.TextSurface
import su.levenetc.android.textsurface.interfaces.IEndListener
import su.levenetc.android.textsurface.interfaces.ISurfaceAnimation
import su.levenetc.android.textsurface.interfaces.ITextEffect
import su.levenetc.android.textsurface.utils.Utils.addEndListener

/**
 * Created by Eugene Levenetc.
 */
class ShapeReveal private constructor(override val text: Text, override var duration: Long, private val revealShape: IRevealShape, private val hideOnEnd: Boolean) : ITextEffect, AnimatorUpdateListener {

    init {
        revealShape.setText(text)
    }

    override lateinit var textSurface: TextSurface
    private var animator: ValueAnimator? = null
    override fun apply(canvas: Canvas, textValue: String, x: Float, y: Float, paint: Paint) {
        revealShape.clip(canvas, textValue, x, y, paint)
    }

    override fun setInitValues(text: Text) {
        if (revealShape.isToShow) text.setAlpha(0)
    }

    override fun onStart() {
        text.addEffect(this)
    }

    override fun start(listener: IEndListener) {
        text.setAlpha(255)
        revealShape.textSurface = textSurface
        animator = revealShape.getAnimator()
        animator!!.interpolator = FastOutSlowInInterpolator()
        addEndListener(this, animator!!, object : IEndListener {
            override fun onAnimationEnd(animation: ISurfaceAnimation) {
                text.removeEffect(this@ShapeReveal)
                if (hideOnEnd) text.setAlpha(0)
                listener?.onAnimationEnd(this@ShapeReveal)
            }
        })
        animator!!.duration = duration
        animator!!.start()
    }

    override fun cancel() {
        if (animator != null && animator!!.isRunning) {
            animator!!.cancel()
            animator = null
        }
    }

    override fun onAnimationUpdate(animation: ValueAnimator) {
        textSurface!!.invalidate()
    }


    companion object {
        @kotlin.jvm.JvmStatic
        fun create(text: Text, duration: Long, revealShape: IRevealShape, hideOnEnd: Boolean): ShapeReveal {
            return ShapeReveal(text, duration, revealShape, hideOnEnd)
        }
    }
}